File: src/lib/api/sources/live-check.ts
import { invoke } from "@tauri-apps/api/core";

import { decodeInstalledSource, type InstalledSource } from "./installed";
import type { Diagnostics, JsonObject } from "./profiles";

export type CheckReportKind = "source_live_check";
export type CheckReportSubjectType = "source";
export type CheckReportResult = "passed" | "failed";
export type CheckReportSubject = { type: CheckReportSubjectType; key: string };
export type CheckFingerprint = {
  kind: string;
  sha256?: string;
  reference?: string;
};
export type CheckReport = {
  schemaVersion: 1;
  kind: CheckReportKind;
  subject: CheckReportSubject;
  checkedAt: string;
  logicVersion: string;
  result: CheckReportResult;
  fingerprints: CheckFingerprint[];
  diagnostics: Diagnostics;
  details: JsonObject;
};
export type CheckReportFreshnessState = "fresh" | "stale";
export type CheckReportStaleReason =
  | "logic_version_changed"
  | "missing_report_fingerprint"
  | "changed_fingerprint_sha256"
  | "unexpected_report_fingerprint";
export type CheckReportStaleDetail = {
  kind: string;
  reference?: string;
  reason: CheckReportStaleReason;
  expectedSha256?: string;
  actualSha256?: string;
  expectedValue?: string;
  actualValue?: string;
};
export type CheckReportFreshness = {
  state: CheckReportFreshnessState;
  staleFingerprints: CheckReportStaleDetail[];
};
export type SourceLiveCheckReportStatus = {
  state: "fresh" | "stale" | "unknown";
  report: CheckReport | null;
  freshness: CheckReportFreshness | null;
};
export type SourceLiveCheckRunOutcome = { report: CheckReport };
export type SourceLiveCheckAdmissionOutcome =
  | { type: "checked"; report: CheckReport }
  | { type: "activated"; report: CheckReport; source: InstalledSource };
export type SourceLiveCheckErrorKind =
  | "invalid_key"
  | "not_found"
  | "built_in"
  | "invalid_lifecycle"
  | "stale_generation"
  | "limit_exceeded"
  | "storage"
  | "check";

export class SourceLiveCheckCommandError extends Error {
  constructor(
    public readonly kind: SourceLiveCheckErrorKind,
    message: string,
  ) {
    super(message);
    this.name = "SourceLiveCheckCommandError";
  }
}

export async function checkSource(sourceKey: string): Promise<SourceLiveCheckRunOutcome> {
  return decodeRunOutcome(await invokeLiveCheck("check_source", sourceKey));
}

export async function checkAndActivateSource(
  sourceKey: string,
): Promise<SourceLiveCheckAdmissionOutcome> {
  return decodeAdmissionOutcome(
    await invokeLiveCheck("check_and_activate_source", sourceKey),
  );
}

export async function getSourceLiveCheckReportStatus(
  sourceKey: string,
): Promise<SourceLiveCheckReportStatus> {
  return decodeSourceLiveCheckReportStatus(
    await invokeLiveCheck("get_source_live_check_report_status", sourceKey),
  );
}

async function invokeLiveCheck(command: string, sourceKey: string): Promise<unknown> {
  try {
    return await invoke<unknown>(command, { sourceKey });
  } catch (error) {
    throw decodeSourceLiveCheckError(error);
  }
}

export function decodeSourceLiveCheckError(error: unknown): Error {
  if (
    isRecord(error) &&
    [
      "invalid_key",
      "not_found",
      "built_in",
      "invalid_lifecycle",
      "stale_generation",
      "limit_exceeded",
      "storage",
      "check",
    ].includes(String(error.kind)) &&
    typeof error.message === "string"
  ) {
    return new SourceLiveCheckCommandError(
      error.kind as SourceLiveCheckErrorKind,
      error.message,
    );
  }
  return error instanceof Error ? error : new Error(String(error));
}

export function decodeRunOutcome(value: unknown): SourceLiveCheckRunOutcome {
  if (!isRecord(value)) invalid();
  return { report: decodeReport(value.report) };
}

export function decodeAdmissionOutcome(
  value: unknown,
): SourceLiveCheckAdmissionOutcome {
  if (!isRecord(value)) invalid();
  const report = decodeReport(value.report);
  if (value.type === "checked" && report.result === "failed") {
    return { type: "checked", report };
  }
  if (value.type === "activated" && report.result === "passed") {
    const source = decodeInstalledSource(value.source);
    if (source.document.status !== "active") return invalid();
    return { type: "activated", report, source };
  }
  return invalid();
}

export function decodeSourceLiveCheckReportStatus(
  value: unknown,
): SourceLiveCheckReportStatus {
  if (!isRecord(value) || !["fresh", "stale", "unknown"].includes(String(value.state))) {
    return invalid();
  }
  if (value.state === "unknown") {
    if (value.report != null || value.freshness != null) return invalid();
    return { state: "unknown", report: null, freshness: null };
  }
  const report = decodeReport(value.report);
  const freshness = decodeFreshness(value.freshness);
  if (freshness.state !== value.state) return invalid();
  return { state: value.state, report, freshness } as SourceLiveCheckReportStatus;
}

function decodeReport(value: unknown): CheckReport {
  if (
    !isRecord(value) ||
    value.schemaVersion !== 1 ||
    value.kind !== "source_live_check" ||
    !isRecord(value.subject) ||
    value.subject.type !== "source" ||
    typeof value.subject.key !== "string" ||
    typeof value.checkedAt !== "string" ||
    typeof value.logicVersion !== "string" ||
    (value.result !== "passed" && value.result !== "failed") ||
    !Array.isArray(value.fingerprints) ||
    !Array.isArray(value.diagnostics) ||
    !isRecord(value.details)
  ) return invalid();
  for (const fingerprint of value.fingerprints) {
    if (
      !isRecord(fingerprint) ||
      typeof fingerprint.kind !== "string" ||
      (fingerprint.reference !== undefined && typeof fingerprint.reference !== "string") ||
      (fingerprint.sha256 !== undefined &&
        (typeof fingerprint.sha256 !== "string" ||
          !/^[a-fA-F0-9]{64}$/.test(fingerprint.sha256)))
    ) return invalid();
  }
  for (const diagnostic of value.diagnostics) {
    if (
      !isRecord(diagnostic) ||
      ![
        "schema",
        "registry",
        "compiler",
        "runtime",
        "detection",
        "source_validation",
      ].includes(String(diagnostic.category)) ||
      typeof diagnostic.code !== "string" ||
      typeof diagnostic.message !== "string" ||
      !["info", "warning", "error"].includes(String(diagnostic.severity)) ||
      typeof diagnostic.path !== "string" ||
      (diagnostic.strategyKey !== undefined &&
        typeof diagnostic.strategyKey !== "string") ||
      (diagnostic.details !== undefined && !isJsonValue(diagnostic.details))
    ) return invalid();
  }
  return value as CheckReport;
}

function decodeFreshness(value: unknown): CheckReportFreshness {
  if (
    !isRecord(value) ||
    (value.state !== "fresh" && value.state !== "stale") ||
    !Array.isArray(value.staleFingerprints)
  ) return invalid();
  if (value.state === "fresh" && value.staleFingerprints.length !== 0) return invalid();
  if (value.state === "stale" && value.staleFingerprints.length === 0) return invalid();
  for (const detail of value.staleFingerprints) {
    if (
      !isRecord(detail) ||
      typeof detail.kind !== "string" ||
      ![
        "logic_version_changed",
        "missing_report_fingerprint",
        "changed_fingerprint_sha256",
        "unexpected_report_fingerprint",
      ].includes(String(detail.reason)) ||
      [
        detail.reference,
        detail.expectedSha256,
        detail.actualSha256,
        detail.expectedValue,
        detail.actualValue,
      ].some((item) => item !== undefined && typeof item !== "string")
    ) return invalid();
  }
  return value as CheckReportFreshness;
}

function isJsonValue(value: unknown): boolean {
  if (value === null || ["string", "number", "boolean"].includes(typeof value)) return true;
  if (Array.isArray(value)) return value.every(isJsonValue);
  return isRecord(value) && Object.values(value).every(isJsonValue);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function invalid(): never {
  throw new Error("invalid Source Live Check transport");
}

File: src-tauri/crates/sources/src/live_check.rs
//! Bounded Source Live Check execution, evidence, freshness, and admission.

use std::{
    fmt,
    io::ErrorKind as IoErrorKind,
    path::PathBuf,
    sync::Arc,
    time::{SystemTime, UNIX_EPOCH},
};

use serde::Serialize;
use source_profile_dsl::definition::{Diagnostic, DiagnosticCategory, DiagnosticSeverity};
use source_profile_dsl::execution::{
    BoxedBrowserAcquisitionFuture, BrowserAcquisition, BrowserAcquisitionRequest,
    ProfileHttpClient, ProfileHttpError, ProfileHttpRequest, ProfileHttpResponse,
    RuntimeCancellation, RuntimeExecutionContext,
};

use crate::installed::{self, Store};

mod execution;
mod fingerprint;
mod fingerprints;
mod freshness;
mod persistence;
mod report;

#[cfg(test)]
mod fingerprint_tests;
#[cfg(test)]
mod freshness_tests;

pub use fingerprint::CheckFingerprint as Fingerprint;
pub use freshness::{
    CheckReportFreshness as Freshness, CheckReportFreshnessState as FreshnessState,
    CheckReportStaleDetail as StaleDetail, CheckReportStaleReason as StaleReason,
};
pub use report::{
    CheckReport as Report, CheckReportKind as ReportKind, CheckReportResult as ReportResult,
    CheckReportSubject as ReportSubject, CheckReportSubjectType as ReportSubjectType,
    CHECK_REPORT_SCHEMA_VERSION,
};

pub const LOGIC_VERSION: &str = "source-live-check/v2";

pub trait Clock: Send + Sync {
    fn now(&self) -> String;
}

#[derive(Clone, Copy, Debug, Default)]
pub struct SystemClock;

impl Clock for SystemClock {
    fn now(&self) -> String {
        let seconds = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs() as i64;
        let days = seconds.div_euclid(86_400);
        let seconds_of_day = seconds.rem_euclid(86_400);
        let z = days + 719_468;
        let era = if z >= 0 { z } else { z - 146_096 } / 146_097;
        let day_of_era = z - era * 146_097;
        let year_of_era =
            (day_of_era - day_of_era / 1_460 + day_of_era / 36_524 - day_of_era / 146_096) / 365;
        let mut year = year_of_era + era * 400;
        let day_of_year = day_of_era - (365 * year_of_era + year_of_era / 4 - year_of_era / 100);
        let month_parameter = (5 * day_of_year + 2) / 153;
        let day = day_of_year - (153 * month_parameter + 2) / 5 + 1;
        let month = month_parameter + if month_parameter < 10 { 3 } else { -9 };
        year += i64::from(month <= 2);
        let hour = seconds_of_day / 3_600;
        let minute = (seconds_of_day % 3_600) / 60;
        let second = seconds_of_day % 60;
        format!("{year:04}-{month:02}-{day:02}T{hour:02}:{minute:02}:{second:02}Z")
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ReportState {
    Fresh,
    Stale,
    Unknown,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Status {
    pub state: ReportState,
    pub report: Option<Report>,
    pub freshness: Option<Freshness>,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RunOutcome {
    pub report: Report,
}

#[derive(Clone, Debug, Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum AdmissionOutcome {
    Checked {
        report: Report,
    },
    Activated {
        report: Report,
        source: installed::SourceView,
    },
}

#[derive(Clone, Default)]
pub struct Context {
    pub cancellation: Option<Arc<dyn RuntimeCancellation>>,
}

#[derive(Clone)]
pub struct Operation {
    app_data_dir: PathBuf,
    installed: Store,
    http: Arc<dyn ProfileHttpClient + Send + Sync>,
    browser: Arc<dyn BrowserAcquisition>,
    clock: Arc<dyn Clock>,
}

impl Operation {
    pub fn new(
        installed: Store,
        http: Arc<dyn ProfileHttpClient + Send + Sync>,
        browser: Arc<dyn BrowserAcquisition>,
        clock: Arc<dyn Clock>,
    ) -> Self {
        Self {
            app_data_dir: installed.app_data_dir().to_path_buf(),
            installed,
            http,
            browser,
            clock,
        }
    }

    pub async fn status(&self, source_key: &str) -> Result<Status, Error> {
        validate_key(source_key)?;
        let installed = self.installed.clone();
        let source_key = source_key.to_string();
        let app_data_dir = self.app_data_dir.clone();
        tokio::task::spawn_blocking(move || {
            let snapshot = installed.snapshot().map_err(Error::installed)?;
            if snapshot.source(&source_key).is_none() {
                return Err(Error::not_found(&source_key));
            }
            let path = persistence::source_live_check_report_path(app_data_dir, &source_key);
            match persistence::read_latest_check_report(path) {
                Ok(report) => {
                    let current =
                        execution::prepare(&snapshot, &source_key).map_err(Error::check)?;
                    let freshness = freshness::evaluate_check_report_freshness(
                        &report,
                        LOGIC_VERSION,
                        &current.fingerprints,
                    );
                    let state = match freshness.state {
                        freshness::CheckReportFreshnessState::Fresh => ReportState::Fresh,
                        freshness::CheckReportFreshnessState::Stale => ReportState::Stale,
                    };
                    Ok(Status {
                        state,
                        report: Some(report),
                        freshness: Some(freshness),
                    })
                }
                Err(persistence::CheckReportPersistenceError::Io(error))
                    if error.kind() == IoErrorKind::NotFound =>
                {
                    Ok(Status {
                        state: ReportState::Unknown,
                        report: None,
                        freshness: None,
                    })
                }
                Err(error) => Err(Error::storage(error.to_string())),
            }
        })
        .await
        .map_err(|error| Error::storage(error.to_string()))?
    }

    pub async fn run(&self, source_key: &str, context: Context) -> Result<RunOutcome, Error> {
        validate_key(source_key)?;
        let installed = self.installed.clone();
        let snapshot = tokio::task::spawn_blocking(move || installed.snapshot())
            .await
            .map_err(|error| Error::storage(error.to_string()))?
            .map_err(Error::installed)?;
        if snapshot.source(source_key).is_none() {
            return Err(Error::not_found(source_key));
        }
        let execution_context = execution::ExecutionContext {
            checked_at: self.clock.now(),
            cancellation: context.cancellation.as_deref(),
        };
        let http = SharedHttp(Arc::clone(&self.http));
        let browser = SharedBrowser(Arc::clone(&self.browser));
        let mut report = execution::build_report(
            &snapshot,
            source_key,
            &http,
            &http,
            &browser,
            &execution_context,
        )
        .await
        .map_err(Error::check)?;
        mark_cancelled_if_observed(&mut report, context.cancellation.as_deref());
        self.persist(&report).await?;
        Ok(RunOutcome { report })
    }

    pub async fn check_and_activate(
        &self,
        source_key: &str,
        context: Context,
    ) -> Result<AdmissionOutcome, Error> {
        validate_key(source_key)?;
        let installed = self.installed.clone();
        let snapshot = tokio::task::spawn_blocking(move || installed.snapshot())
            .await
            .map_err(|error| Error::storage(error.to_string()))?
            .map_err(Error::installed)?;
        let prepared = snapshot
            .source(source_key)
            .ok_or_else(|| Error::not_found(source_key))?;
        if prepared.origin() != installed::Origin::Custom {
            return Err(Error::new(
                ErrorKind::BuiltIn,
                format!("built-in Source `{source_key}` cannot be mutated"),
            ));
        }
        if prepared.document().status == installed::SourceStatus::Active {
            return Err(Error::new(
                ErrorKind::InvalidLifecycle,
                "an active Source cannot be check-and-activated",
            ));
        }
        let generation = prepared.generation().clone();
        let execution_context = execution::ExecutionContext {
            checked_at: self.clock.now(),
            cancellation: context.cancellation.as_deref(),
        };
        let http = SharedHttp(Arc::clone(&self.http));
        let browser = SharedBrowser(Arc::clone(&self.browser));
        let mut report = execution::build_report(
            &snapshot,
            source_key,
            &http,
            &http,
            &browser,
            &execution_context,
        )
        .await
        .map_err(Error::check)?;
        mark_cancelled_if_observed(&mut report, context.cancellation.as_deref());
        self.persist(&report).await?;
        if report.result != ReportResult::Passed {
            return Ok(AdmissionOutcome::Checked { report });
        }
        let installed = self.installed.clone();
        let source_key = source_key.to_string();
        let cancellation = context.cancellation.clone();
        let source = tokio::task::spawn_blocking(move || {
            installed.admit_checked_if(&source_key, &generation, || {
                !cancellation
                    .as_deref()
                    .is_some_and(RuntimeCancellation::is_cancelled)
            })
        })
        .await
        .map_err(|error| Error::storage(error.to_string()))?
        .map_err(Error::installed)?;
        let Some(source) = source else {
            mark_cancelled(&mut report);
            self.persist(&report).await?;
            return Ok(AdmissionOutcome::Checked { report });
        };
        Ok(AdmissionOutcome::Activated { report, source })
    }

    async fn persist(&self, report: &Report) -> Result<(), Error> {
        let app_data_dir = self.app_data_dir.clone();
        let report = report.clone();
        tokio::task::spawn_blocking(move || {
            persistence::persist_latest_check_report(app_data_dir, &report)
        })
        .await
        .map_err(|error| Error::storage(error.to_string()))?
        .map(|_| ())
        .map_err(|error| Error::storage(error.to_string()))
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ErrorKind {
    InvalidKey,
    NotFound,
    BuiltIn,
    InvalidLifecycle,
    StaleGeneration,
    LimitExceeded,
    Storage,
    Check,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Error {
    pub kind: ErrorKind,
    pub message: String,
}

impl Error {
    fn new(kind: ErrorKind, message: impl Into<String>) -> Self {
        Self {
            kind,
            message: message.into(),
        }
    }
    fn not_found(key: &str) -> Self {
        Self::new(ErrorKind::NotFound, format!("Source `{key}` was not found"))
    }
    fn storage(message: impl Into<String>) -> Self {
        Self::new(ErrorKind::Storage, message)
    }
    fn check(message: impl Into<String>) -> Self {
        Self::new(ErrorKind::Check, message)
    }
    fn installed(error: installed::Error) -> Self {
        let kind = match error.kind {
            installed::ErrorKind::InvalidKey => ErrorKind::InvalidKey,
            installed::ErrorKind::NotFound => ErrorKind::NotFound,
            installed::ErrorKind::BuiltIn => ErrorKind::BuiltIn,
            installed::ErrorKind::InvalidLifecycle => ErrorKind::InvalidLifecycle,
            installed::ErrorKind::GenerationMismatch => ErrorKind::StaleGeneration,
            installed::ErrorKind::LimitExceeded => ErrorKind::LimitExceeded,
            installed::ErrorKind::Storage | installed::ErrorKind::Load => ErrorKind::Storage,
            installed::ErrorKind::Duplicate | installed::ErrorKind::InvalidInput => {
                unreachable!("Source Live Check does not perform authored Source mutations")
            }
        };
        Self::new(kind, error.message)
    }
}

impl fmt::Display for Error {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(&self.message)
    }
}

impl std::error::Error for Error {}

fn mark_cancelled_if_observed(report: &mut Report, cancellation: Option<&dyn RuntimeCancellation>) {
    if cancellation.is_some_and(RuntimeCancellation::is_cancelled) {
        mark_cancelled(report);
    }
}

fn mark_cancelled(report: &mut Report) {
    if report.result == ReportResult::Failed
        && report
            .diagnostics
            .iter()
            .any(|diagnostic| diagnostic.code == "runtime_execution_cancelled")
    {
        return;
    }
    report.result = ReportResult::Failed;
    report.details.insert(
        "liveCheckState".to_string(),
        serde_json::Value::String("live_check_failed".to_string()),
    );
    report.diagnostics.push(Diagnostic {
        category: DiagnosticCategory::Runtime,
        code: "runtime_execution_cancelled".to_string(),
        message: "Source Live Check was cancelled before admission".to_string(),
        severity: DiagnosticSeverity::Error,
        path: "".to_string(),
        strategy_key: None,
        details: None,
    });
}

#[derive(Clone)]
struct SharedHttp(Arc<dyn ProfileHttpClient + Send + Sync>);

impl ProfileHttpClient for SharedHttp {
    fn fetch<'a>(
        &'a self,
        request: ProfileHttpRequest,
        context: RuntimeExecutionContext<'a>,
    ) -> std::pin::Pin<
        Box<
            dyn std::future::Future<Output = Result<ProfileHttpResponse, ProfileHttpError>>
                + Send
                + 'a,
        >,
    > {
        self.0.fetch(request, context)
    }
}

#[derive(Clone)]
struct SharedBrowser(Arc<dyn BrowserAcquisition>);

impl BrowserAcquisition for SharedBrowser {
    fn acquire<'a>(
        &'a self,
        request: BrowserAcquisitionRequest<'a>,
    ) -> BoxedBrowserAcquisitionFuture<'a> {
        self.0.acquire(request)
    }
}

fn validate_key(key: &str) -> Result<(), Error> {
    let mut characters = key.chars();
    if characters
        .next()
        .is_some_and(|character| character.is_ascii_lowercase() || character.is_ascii_digit())
        && characters.all(|character| {
            character.is_ascii_lowercase() || character.is_ascii_digit() || character == '_'
        })
    {
        Ok(())
    } else {
        Err(Error::new(
            ErrorKind::InvalidKey,
            format!("invalid Source key `{key}`"),
        ))
    }
}
